### ISSUE TITLE
Return mean Y given W, X as well as difference

### ISSUE THREAD

**Issue by gannawag** [grf-labs/grf#295]
Currently the predict.causal_forest function returns the predicted treatment effect given the X values for each observation in the new data. 
Is there a way to also return the mean (over all trees) of the leaf-specific mean of Y for both treatment and control groups? 
A previous issue mentions knowing which leaves each observation lands in, it would be easy to back out the leaf-specific means of Y from there.

**Comment by susanathey** [comment]
I agree this would be useful, would be something that researchers might want to plot.  We could also put in the standard errors.  Perhaps this could be turned on as an option; need to think if we are already saving enough information to do this easily.

**Comment by jtibshirani** [comment]
It would also be nice if these quantities were available on a per-tree basisl, so they would be available in the tree visualizations that @Buyannemekh is working on in #286.

**Comment by gannawag** [comment]
Getting closer with this. Is there an easy way to reference the non-centered outcome and treatment variables in the instrumental forest? `observations.get(Observations::OUTCOME, sample)` in `InstrumentalPredictionStrategy.cpp` seems to only reference the centered versions.

**Comment by swager** [comment]
Yup -- like you mention, @gannawag, all the quantities stored by `causal_forest` reference only the centered outcomes and treatments. Moreover, we always assume a continuous treatment (because even if `W` is binary, `W - W.hat` is not binary). The upshot is that getting anything that can be written in terms of these quantities is easy, e.g., the leaf-wise average of `(Y - Y.hat) * (W - W.hat)`, but anything that depends on the uncentered data isn't directly accessible.

If the underlying goal is to get estimates of `mu_w(x) = E[Y(w) | X=x]`, then we recommend using
```
hat{mu}_w(x) = hat{m}(x) + (w - hat{e}(x)) * hat{tau}(x)
```
where `hat{tau}(x)` is the CATE estimate and `hat{m}(x)` and `hat{e}(x)` are the first stages estimates for `E[Y|X=x]` and `E[W|X=x]` respectively (these are the quantities we use for centering). This is what we do in the `average_treatment_effect` function.

More broadly, we expose the forest weights `alpha_i(x)` via the function `get_sample_weights`, so you may be able to use that for richer analyses.

**Comment by gannawag** [comment]
What about this approach:

1. Edit `causal_forest.R` to define the data used in training as `data <- create_data_matrices(X, Y.centered, W.centered, Y)`

2. Edit `InstrumentalPredictionStrategy.cpp` to return (along with current returned values) the mean of Y (uncentered) in each node by treatment status. This was easy enough, I just defined a new function that returns new values I created in the `precompute_prediction_values()` function. 

I can post my (hack-job) code if it would be useful. My approach doesn't change how the trees are defined, it just gives information about the observations in each leaf. 

I also extended this idea a bit to allow me to find the raw uncentered difference in treatment and control means for other variables as well. The idea here is show other outcomes affected by the treatment but that on which we don't want to partition the data, or as a placebo to show that outcomes that should not be affected by the treatment indeed aren't.

**Comment by swager** [comment]
This kind of placebo test seems like a very nice thing to do. Would the strategy using the `get_sample_weights` command not work, though? You can do this kind of thing:
```
n = 500; n.test = 50; p = 10
X = matrix(rnorm(n*p), n, p)
Y = X[,2] + pmin(X[,3], 0) + rnorm(n)
rf = regression_forest(X, Y)

X.test = matrix(rnorm(n.test*p), n.test, p)
alpha.test = get_sample_weights(rf, X.test)
mu.hat.direct = predict(rf, X.test)$predictions
mu.hat.alpha = alpha.test %*% Y
max(abs(mu.hat.direct - mu.hat.alpha))
```
Then you can replace `Y` with a placebo in `alpha.test %*% Y`. The `get_sample_weights` function can also be used to similarly get the weighting function of a causal (or other) forest.

**Comment by gannawag** [comment]
This works exactly as I need it to (thanks!), except for one thing: returning an n-by-n matrix of weights takes a very long time when I have a lot of data. This seems like it would most often be a problem when using oob predictions or placebo predictions on the training data set. I do agree that once the weights matrix is available, there are a ton of possibilities.

**Comment by jtibshirani** [comment]
Hi @gannawag, I'm sorry for the slow response. We've also noticed that retrieving weights is quite slow and plan to address this as part of #230, which aims to parallelize this computation.

Catching up with @swager, we've decided not to add dedicated support for retrieving these summary statistics, as the quantities that are useful will really depend on the specific analysis. Instead we'll focus on making the `get_sample_weights` call more performant.
